proxy/grpcproxy/lease.go

```diff
@@ -46,42 +94,209 @@ func (lp *leaseProxy) LeaseTimeToLive(ctx context.Context, rr *pb.LeaseTimeToLiv
+	// lrpCh is a channel to receive lease keepalive response from etcd backend
+	lrpCh chan *pb.LeaseKeepAliveResponse
+
+	ctx context.Context
+}
+
+func (lps *leaseProxyStream) recvLoop() error {
+	defer func() {
+		fmt.Printf("leaseProxyStream %v closed \n", lps.stream)
+	}()
+	for lps.ctx.Err() == nil {
+		fmt.Printf("leaseProxyStream %v receiving msg...\n", lps.stream)
+		rr, err := lps.stream.Recv()
+		fmt.Printf("leaseProxyStream %v received msg %v \n", lps.stream, rr)
+		if err == io.EOF {
+			return nil
 		}
-		err = stream.Send(rr)
 		if err != nil {
 			return err
 		}
+		lps.serveKeepAliveRequest(rr)
+	}
+	return lps.ctx.Err()
+}
+
+func (lps *leaseProxyStream) serveKeepAliveRequest(rr *pb.LeaseKeepAliveRequest) {
+	fmt.Printf("leaseProxyStream.serveKeepAliveRequest() req %v ...\n", rr)
+	lps.mu.Lock()
+	defer lps.mu.Unlock()
+	msgsCh, ok := lps.keepAliveLeases[rr.ID]
+	if !ok {
+		msgsCh = make(chan struct{}, 16)
+		lps.keepAliveLeases[rr.ID] = msgsCh
+		lps.wg.Add(1)
+		go func() {
+			defer lps.wg.Done()
+			lps.listeningFromServer(rr.ID)
+		}()
+	}
+	select {
+	case msgsCh <- struct{}{}:
+		fmt.Printf("leaseProxyStream.serveKeepAliveRequest() put items in msgsCh \n")
+	default:
+	}
+	fmt.Printf("leaseProxyStream.serveKeepAliveRequest() req %v done \n", rr)
+}
+
+func (lps *leaseProxyStream) listeningFromServer(leaseID int64) {
+	fmt.Printf("leaseProxyStream.listeningFromServer() %v ...\n", leaseID)
+	var (
+		msgsCh <-chan struct{}
+		ticker <-chan time.Time
+	)
+	cctx, ccancel := context.WithCancel(lps.ctx)
+	defer func() {
+		// clean up
+		ccancel()
+		lps.mu.Lock()
+		delete(lps.keepAliveLeases, leaseID)
+		lps.mu.Unlock()
+		fmt.Printf("leaseProxyStream.listeningFromServer() %v done \n", leaseID)
+	}()
+	rpch, err := lps.client.KeepAlive(cctx, clientv3.LeaseID(leaseID))
+	if err != nil {
+		fmt.Printf("leaseProxyStream.listeningFromServer() %v creating KeepAlive() stream error %v \n", leaseID, err)
+		return
+	}
+	lps.mu.RLock()
+	msgsCh = lps.keepAliveLeases[leaseID]
+	lps.mu.RUnlock()
+
+	for {
+		select {
+		case <-cctx.Done():
+			fmt.Printf("leaseProxyStream.listeningFromServer() %v context is done \n", leaseID)
+			return
+		case <-ticker:
+			fmt.Printf("leaseProxyStream.listeningFromServer() %v timeout \n", leaseID)
+
+		case <-msgsCh:
+			fmt.Printf("leaseProxyStream.listeningFromServer() %v msgsCh is not empty \n", leaseID)
+			rp, ok := <-rpch
+			fmt.Printf("leaseProxyStream.listeningFromServer() received KeepAlive %v from server \n", rp)
+			if !ok {
+				fmt.Printf("leaseProxyStream.listeningFromServer() %v rpch is closed \n", leaseID)
+				// rpch is closed probably due to proxy_to_etcd client is unable to receive keepAlive response
+				return
+			}
+			r := &pb.LeaseKeepAliveResponse{
+				Header: rp.ResponseHeader,
+				ID:     int64(rp.ID),
+				TTL:    rp.TTL,
+			}
+			ticker = time.After(time.Duration(r.TTL) * time.Second)
+			select {
+			case lps.lrpCh <- r:
+			default:
```

Comment:
That's not how the lease grpc stream protocol works. See https://github.com/coreos/etcd/blob/master/etcdserver/api/v3rpc/lease.go#L63 -- where does it drop requests on the floor? 

It's not best-effort; it will not drop requests without sending a response. If it doesn't eventually respond to every request, that should mean the stream is lost.